lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs

```diff
@@ -0,0 +1,228 @@
+    data Prologue s :: Type
+    -- | Addresses that were collected during discovery on the blockchain.
+    data Discoveries s :: Type
+
+    -- | Isomorphism between the address book type 's'
+    -- and its two components.
+    addressIso :: Iso' s (Prologue s, Discoveries s)
+
+{-------------------------------------------------------------------------------
+    Sequential address book
+-------------------------------------------------------------------------------}
+-- | Sequential list of addresses
+newtype SeqAddressList (c :: Role) = SeqAddressList [(W.Address, W.AddressState)]
+
+-- piggy-back on SeqState existing instance, to simulate the same behavior.
+instance AddressBookIso (Seq.SeqState n k)
+    => AddressBookIso (Seq.SeqAnyState n k p)
+  where
+    data Prologue (Seq.SeqAnyState n k p) = PS (Prologue (Seq.SeqState n k))
+    data Discoveries (Seq.SeqAnyState n k p) = DS (Discoveries (Seq.SeqState n k))
+
+    addressIso = withIso addressIso $ \from to ->
+        let from2 st = let (a,b) = from $ Seq.innerState st in (PS a, DS b)
+            to2 (PS a, DS b) = Seq.SeqAnyState $ to (a,b)
+        in  iso from2 to2
+
+-- | Isomorphism for sequential address book.
+instance 
+    ( MkKeyFingerprint key (Proxy n, key 'AddressK XPub)
+    , GetPurpose key
+    , SoftDerivation key
+    , PaymentAddress n key
+    , Typeable n
+    , (key == SharedKey) ~ 'False
+    ) => AddressBookIso (Seq.SeqState n key)
+  where
+    data Prologue (Seq.SeqState n key)
+        = SeqPrologue (Seq.SeqState n key)
+        -- Trick: We keep the type, but we empty the discovered addresses
+    data Discoveries (Seq.SeqState n key)
+        = SeqDiscoveries
+            (SeqAddressList 'UtxoInternal)
+            (SeqAddressList 'UtxoExternal)
+
+    addressIso = iso from to
+      where
+        from (Seq.SeqState int ext a b c) =
+            ( SeqPrologue $ Seq.SeqState (emptyPool int) (emptyPool ext) a b c
+            , SeqDiscoveries (toDiscoveries @n int) (toDiscoveries @n ext)
+            )
+        to (SeqPrologue (Seq.SeqState int ext a b c), SeqDiscoveries ints exts)
+          = Seq.SeqState (fromDiscoveries @n int ints) (fromDiscoveries @n ext exts) a b c
+
+-- | Extract the discovered addresses from an address pool.
+toDiscoveries
+    :: forall (n :: NetworkDiscriminant) (c :: Role) key.
+    ( GetPurpose key
+    , PaymentAddress n key
+    , Typeable c
+    ) => Seq.AddressPool c key -> SeqAddressList c
+toDiscoveries pool = SeqAddressList
+    [ (a,st) | (a,st,_) <- Seq.addresses (liftPaymentAddress @n) pool ]
+
+-- | Variant of 'toDiscoveries' that can be used with 'SharedKey'
+toDiscoveriesShared
+    :: forall (n :: NetworkDiscriminant) (c :: Role) key.
+    ( GetPurpose key
+    , Typeable c
+    , Typeable n
+    , key ~ SharedKey
+    ) => Seq.AddressPool c key -> SeqAddressList c
+toDiscoveriesShared pool = SeqAddressList
+    [ (a,st) | (a,st,_) <- Seq.addresses (Shared.liftPaymentAddress @n) pool ]
```

Comment:
the same comment as before... if we want to cover all possible addresses having script hashes in credentials this could not be 1-to-1 mapping anymore also here